tools/clang/test/HLSL/atomics-on-bitfields.hlsl

```diff
@@ -0,0 +1,30 @@
+// RUN: %clang_cc1 -HV 2021 -fsyntax-only -ffreestanding -verify %s
+
+// Ensure that atomic operations fail when used with bitfields
+// Use structured and typed buffers as the resources that can use structs
+// and also both binary op and exchange atomic ops as either difference
+// can cause the compiler to take different paths
+
+struct TexCoords {
+  uint s : 8;
+  uint t : 8;
+  uint r : 8;
+  uint q : 8;
+};
+
+RWBuffer<TexCoords> buf;
+RWStructuredBuffer<TexCoords> str;
+groupshared TexCoords gs;
+
+[numthreads(8,8,1)]
+void main( uint2 tid : SV_DispatchThreadID) {
+
+  InterlockedOr(buf[tid.y].q, 2);                           /* expected-error {{no matching function for call to 'InterlockedOr'}} expected-note {{candidate function not viable: no known conversion from 'uint' to 'unsigned int &' for 1st argument}} expected-note {{candidate function not viable: no known conversion from 'uint' to 'unsigned long long &' for 1st argument}} */
+  InterlockedCompareStore(buf[tid.y].q, 3, 1);              /* expected-error {{no matching function for call to 'InterlockedCompareStore'}} expected-note {{candidate function not viable: no known conversion from 'uint' to 'unsigned int &' for 1st argument}} expected-note {{candidate function not viable: no known conversion from 'uint' to 'unsigned long long &' for 1st argument}} */
+
+  InterlockedOr(str[tid.y].q, 2);                           /* expected-error {{no matching function for call to 'InterlockedOr'}} expected-note {{candidate function not viable: no known conversion from 'uint' to 'unsigned int' for 1st argument}} expected-note {{candidate function not viable: no known conversion from 'uint' to 'unsigned long long' for 1st argument}} */
+  InterlockedCompareStore(str[tid.y].q, 3, 1);              /* expected-error {{no matching function for call to 'InterlockedCompareStore'}} expected-note {{candidate function not viable: no known conversion from 'uint' to 'unsigned int' for 1st argument}} expected-note {{candidate function not viable: no known conversion from 'uint' to 'unsigned long long' for 1st argument}} */
+
+  InterlockedOr(gs.q, 2);                                   /* expected-error {{no matching function for call to 'InterlockedOr'}} expected-note {{candidate function not viable: 1st argument ('__attribute__((address_space(3))) uint') is in address space 3, but parameter must be in address space 0}} expected-note {{candidate function not viable: no known conversion from '__attribute__((address_space(3))) uint' to 'unsigned long long' for 1st argument}} */
```

Comment:
It's a shame about the address space error, since it's misleading.

Reply:
Less so than the previous I'd say. Bitfields just make the compiler do your shifting and casting for you. The result of that operation is a local. Granted, shader authors may not know that. We could consider creating some custom mismatch types at some point. That would be useful in all versions of LLVM unless they've seriously overhauled this system